Performance: `ListViewArray` zero-copyable to `ListArray` optimizations

connortsui20 opened the pull request:
Tracking Issue: https://github.com/vortex-data/vortex/issues/4699
Closes: https://github.com/vortex-data/vortex/issues/5053

This PR introduces a performance optimization for `ListViewArray` by adding an `is_zero_copy_to_list` flag that tracks whether the `ListViewArray` can be (near) constant-time converted to a `ListArray` without copying data.

When this flag is true (indicating sorted offsets with no gaps and no overlaps), conversions can bypass the very expensive rebuild process (which just calls `append_scalar` in a loop). 

Additionally, this PR refactors all `ListViewArray` constructor call sites and compute operations throughout the codebase to properly maintain this new invariant. The only time you can set this flag is via the `new_unchecked` constructor which is unsafe, which means the caller better be careful to know what they are doing.

Some things that are not included in this PR:
- Making conversion _actually_ zero-copy would require implementing logic to always allocate an additional `offsets` slot for an `n+1`th offset in `ListViewArray`. This is because `ListArray` has `n+1` offsets, and from a plain `ListViewArray` we have to reallocate the offsets to make space.
- `reset_offsets` right now is probably more expensive than it could be since it uses the `sub_scalar` compute function. We know in advance that `reset_offsets` should _never_ fail in the non-recursive case.
- Validation of this `is_zctl` flag is **SUPER** expensive, but it can be made a bit less expensive by using a `BitBuffer` to track references + a flag for overlaps.
- It would probably be nice to have an `check_is_zero_copy_to_list` function on `ListViewArray` that sets the flag if it detects it is zero-copyable to a `ListArray`. This would be nice when we are reading from disk or converting from an Arrow `GenericListView`.
- This PR **does not** change the on-disk format, which means we lose this metadata when writing to disk. I think that this flag is somewhat similar to the `is_sorted` statistic, but at the same time I feel like it is not correct to store this as a statistic? Not sure.

connortsui20 committed:
`is_zero_copy_to_list` `ListView` to `List` optimization

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

a10y commented on vortex-array/src/arrays/listview/compute/filter.rs:
```diff
@@ -46,17 +46,24 @@ impl FilterKernel for ListViewVTable {
         let new_sizes = compute::filter(sizes.as_ref(), selection_mask)?;
 
+        // Filters will create gaps of unused elements.
+        let is_zctl = false;
```
can just inline this and the comment to the constructor call

a10y commented on vortex-array/src/arrays/listview/compute/take.rs:
```diff
@@ -64,23 +64,27 @@ impl TakeKernel for ListViewVTable {
 
+        // Take can reorder offsets, create gaps, and may introduce overlaps if the `indices`
+        // contain duplicates.
+        let is_zctl = false;
```
also can inline into constructor call

a10y commented on vortex-array/src/arrays/listview/conversion.rs:
```diff
@@ -12,27 +12,35 @@ use crate::vtable::ValidityHelper;
+    // function might not expect the output `ListViewArray` to have a bunch of leading and trailing
+    // garbage data when they turn it back into a `ListArray`.
+    let list = list
+        .reset_offsets(false)
+        .vortex_expect("TODO(connor)[ListView]: This can't fail");
```
you don't have to do this right now, but i think that `Canonical::compact` should be infallible. I don't think there's ever a valid case right now where it could return an error since we've made validity and other array ops infallilble. Then we can delete this expect

connortsui20 replied:
Yeah I think this should come later (see my updated PR description above) and it can be optimized better

joseph-isaacs commented:
What perf changes did you expect from the benchmarks?

connortsui20 commented:
@joseph-isaacs none since none of our benchmarks really test the conversion between list and listview (we have a test file that didn't finish on develop but finishes in 3-4 sec on this branch). Maybe we can add that as a be regression benchmark?

joseph-isaacs commented on fuzz/src/array/mask.rs:
```diff
@@ -57,13 +57,18 @@ pub fn mask_canonical_array(canonical: Canonical, mask: &Mask) -> VortexResult<A
+                    array.offsets().clone(),
+                    array.sizes().clone(),
+                    new_validity,
+                    array.is_zero_copy_to_list(),
```
This feels leeky here

connortsui20 replied:
Leaky in what way? This is basically identical to the `mask` compute function impl

joseph-isaacs commented on vortex-array/src/arrays/listview/tests/operations.rs:
```diff
@@ -156,9 +155,16 @@ fn test_slice_edge_cases(
+            offsets,
+            sizes,
+            Validity::NonNullable,
+            true, // Is zero-copy to list.
```
Seems like a `with_zero_copy` method would read better here

joseph-isaacs commented on vortex-array/src/arrays/listview/array.rs:
```diff
@@ -345,11 +388,87 @@ where
+
+/// Helper function to validate if the [`ListViewArray`] components are actually zero-copyable to
+/// [`ListArray`](crate::arrays::ListArray).
+fn validate_zctl(
+    elements: &dyn Array,
+    offsets_primitive: PrimitiveArray,
+    sizes_primitive: PrimitiveArray,
+) -> VortexResult<()> {
```
is this tested?

connortsui20 replied:
Yes at every call site where we set is_zctl to true

joseph-isaacs reviewed (commented):
Its nice we can avoid changing serialised array. Its telling that we always have a is_zctl comments everywhere

gatesn commented on vortex-array/src/arrays/listview/array.rs:
```diff
@@ -128,39 +137,61 @@ impl ListViewArray {
         offsets: ArrayRef,
         sizes: ArrayRef,
         validity: Validity,
+        is_zctl: bool,
```
In general, boolean arguments to _any_ function are a no no.

But here, I think this logic should be wrapped up in a `From<ListArray> for ListView` impl, rather than the user ever being able to pass this flag. Even the serde `build` function can check the flag and go via ListArray.

connortsui20 replied:
I think the issue with that here is we need a way to specify something is zero-copyable without  going through a `ListArray` (see the call sites where we change one component and then reconstruct things, like in slice or mask).

I think I like @joseph-isaacs's idea of adding an unsafe builder-like `with_zctl` method, but at the same time I feel that also makes it even harder to reason about the unsafety (harder than it already was) as you now need to worry about two unsafe calls instead of 1. Maybe that's fine?

connortsui20 replied:
I'm going to add an unsafe `with_zctl` method

connortsui20 committed:
use `with_zero_copy_to_list` unsafe method

Instead of taking a boolean flag inside `new_unchecked`, it is better to
use a builder-like unsafe method that adds the flag. This separates the
SAFETY comment out.

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

a10y commented on vortex-array/src/arrays/listview/rebuild.rs:
```diff
@@ -53,121 +66,112 @@ impl ListViewArray {
-        // 3. The validity is unchanged and still matches the array length.
+    /// Rebuilds a [`ListViewArray`] by trimming any unused / unreferenced leading and trailing
+    /// elements, which is defined as a contiguous run of values in the `elements` array that are
+    /// not referecened by any views in the corresponding [`ListViewArray`].
+    fn rebuild_trim_elements(&self) -> ListViewArray {
+        let start = if self.is_zero_copy_to_list() {
+            // If offsets are sorted, then the minimum offset is the first offset.
+            self.offset_at(0)
```
does this hold if the first item is null?

connortsui20 replied:
We have the constraint in `ListArray` where even if there are null elements defined by the validity mask, the offsets still must be sorted. So if the `ListViewArray` came from a `ListArray` that had a null first element, this is still valid.

Also when we validate offsets and sizes we completely ignore nulls since for `ListViewArray` we have the constraint that the views must be valid even if the validity mask specifies the view is null.

See https://arrow.apache.org/docs/format/Columnar.html#listview-layout, specifically:

> Every list-view value, including null values, has to guarantee the following invariants:

This does deserve a comment though

Edit: Added the comment in https://github.com/vortex-data/vortex/pull/5129/commits/258a35a5d8c1a8a4e5147a8a65adfa2af6b41243

a10y reviewed (approved):
Looks good, also ran it again with our broken test case file and perf on the last commit is back to acceptable. Ran the fuzzer for a few minutes and made sure it didn't catch issues either.

Let's ship it
